src/hotspot/share/compiler/compileBroker.cpp

```diff
@@ -2136,6 +2131,12 @@ void CompileBroker::invoke_compiler_on_method(CompileTask* task) {
     DTRACE_METHOD_COMPILE_BEGIN_PROBE(method, compiler_name(task_level));
   }
 
+  if (task->directive()->PrintCompilationOption) {
+    ResourceMark rm;
+    task->print_tty();
+  }
+  elapsedTimer time;
```

Comment:
Shouldn't this timer start at the very top of the method? Printing needs time but it kinda adds to the compilation time, so it's probably not wrong to count that time as well when explicitly using print flags.

Reply:
I've kept it this way because the timer was started after printing prior to this change as well. If a dev is benchmarking compilation time, it is better to not include the timer. Printing adds noise that is not there during real-world runs.

Reply:
Okay, I don't have a strong opinion about it. In the old code, we somehow did it in-between: We printed UL, then with `PrintCompilation`, then started the timer and then printed to `CompilationLog::log()`. So, I'm not sure what the original intention was. Maybe someone else can comment on that, too.

Reply:
`elapsedTimer time;` just declares variable. `TraceTime` uses it (with start/stop) later to accumulate time.
So declaration can be placed anywhere before JVMCI code below.

But I think you should be more flexible here. Consider the case when you hit `assert` at the line 2120. Before this change you will see `PrintCompilation` line, with change - you do not. Of cause in hs_err file you will see it because you did not move `log_compile`.  I understand that you need to check directive. I would suggest to keep global `PrintCompilation` code and add variable to mark that line is already printed. Then check it in you code which checks directive to avoid duplicate.

Reply:
I will move `elapsedTimer` to top. No strong opinion from myself either.

> I would suggest to keep global PrintCompilation code and add variable to mark that line is already printed. Then check it in you code which checks directive to avoid duplicate.

I don't understand this. Doesn't this just print everything?

---

We can do method specific compilation before the assert. This change seems to be working:

```
diff --git a/src/hotspot/share/compiler/compileBroker.cpp b/src/hotspot/share/compiler/compileBroker.cpp
index 8d0a87afc17..181dd28ba45 100644
--- a/src/hotspot/share/compiler/compileBroker.cpp
+++ b/src/hotspot/share/compiler/compileBroker.cpp
@@ -2093,6 +2093,7 @@ CompilerDirectives* DirectivesStack::_bottom = NULL;
 //
 void CompileBroker::invoke_compiler_on_method(CompileTask* task) {
   task->print_ul();
+  elapsedTimer time;

   CompilerThread* thread = CompilerThread::current();
   ResourceMark rm(thread);
@@ -2117,11 +2118,16 @@ void CompileBroker::invoke_compiler_on_method(CompileTask* task) {
     // native.  The NoHandleMark before the transition should catch
     // any cases where this occurs in the future.
     methodHandle method(thread, task->method());
-    assert(!method->is_native(), "no longer compile natives");

     // Look up matching directives
     directive = DirectivesStack::getMatchingDirective(method, comp);
     task->set_directive(directive);
+    if (task->directive()->PrintCompilationOption) {
+      ResourceMark rm;
+      task->print_tty();
+    }
+
+    assert(!method->is_native(), "no longer compile natives");

     // Update compile information when using perfdata.
     if (UsePerfData) {
@@ -2131,12 +2137,6 @@ void CompileBroker::invoke_compiler_on_method(CompileTask* task) {
     DTRACE_METHOD_COMPILE_BEGIN_PROBE(method, compiler_name(task_level));
   }

-  if (task->directive()->PrintCompilationOption) {
-    ResourceMark rm;
-    task->print_tty();
-  }
-  elapsedTimer time;
-
   should_break = directive->BreakAtCompileOption || task->check_break_at_flags();
   if (should_log && !directive->LogOption) {
     should_log = false;
```

Although there is still possible that there is failed assertion before printing, i.e. when calling `CompilerThread::current()`. I feel like its ok to just look at hs_err in these cases.

Reply:
May be we are looking the wrong way. Can we do `task->set_directive(directive)` when we create `task`?
Then you don't need to move PrintCompilation code - `task->directive()` will be available at the beginning of this method.

Reply:
> elapsedTimer time; just declares variable. TraceTime uses it (with start/stop) later to accumulate time.
So declaration can be placed anywhere before JVMCI code below.

Got it, thanks for the clarification - then it does not matter much where we place this declaration.

Reply:
I pushed a change that sets the directive in `CompileTask::initialize()`. To make this change work, I had to remove the `const` from `CompileTask::_directive`. This is because the broker needs a non const version of the directive.

This does not change much. Prior to this PR, the broker would `task->setDirective(directive)`, where `directive` is not a const. `directive` is later passed in to the constructor for `Compilation` as a non const. So although the `CompileTask::_directive` was `const DirectiveSet *`, the directive can still be modified from a different reference.